metals/src/main/scala/scala/meta/internal/metals/BSPErrorHandler.scala

```diff
@@ -0,0 +1,172 @@
+              if (
+                lineBefore.contains(BSPErrorHandler.errorInBsp) &&
+                currentLine.contains(line)
+              ) (currentLine, currentIndex)
+              else (currentLine, matchingLine)
+          }
+      Some(lineNumber)
+    } catch {
+      case NonFatal(_) => None
+    }
+
+  private def gotoLogs(line: Int) = {
+    val pos = new l.Position(line, 0)
+    val location = new l.Location(
+      logsPath.toURI.toString(),
+      new l.Range(pos, pos),
+    )
+    languageClient.metalsExecuteClientCommand(
+      ClientCommands.GotoLocation.toExecuteCommandParams(
+        ClientCommands.WindowLocation(
+          location.getUri(),
+          location.getRange(),
+        )
+      )
+    )
+  }
+}
+
+object BSPErrorHandler {
+  def makeShowMessage(
+      message: String
+  ): Either[l.ShowMessageRequestParams, l.MessageParams] = {
+    if (message.length() <= MESSAGE_MAX_LENGTH) {
+      val messageParams = new l.MessageParams(
+        l.MessageType.Error,
+        s"""|$errorHeader:
+            |$message""".stripMargin,
+      )
+      Right(messageParams)
+    } else {
+      val params = new l.ShowMessageRequestParams()
+      params.setMessage(
+        s"""|$errorHeader
+            |${message.take(MESSAGE_MAX_LENGTH)}...
+            |$gotoLogsToSeeFull""".stripMargin
+      )
+      params.setType(l.MessageType.Error)
+      params.setActions(List(goToLogs).asJava)
+      Left(params)
+    }
+  }
+
+  private val errorHeader = "Encountered an error in the build server:"
+  private val goToLogs = new l.MessageActionItem("Go to logs.")
+  private val gotoLogsToSeeFull = "Go to logs to see the full error"
+  private val errorInBsp = "Error in BSP:"
+
+  private val MESSAGE_MAX_LENGTH = 150
+
+}
+
+class CurrentAndPreviousError(
+    previousError: Option[String] = None,
+    currentError: String = "",
+) {
+  lazy val error: Option[String] = {
+    if (currentError.isEmpty()) None
+    else {
+      val deduplicatedError = {
+        currentError.linesIterator.headOption
+          .map { firstLine =>
+            currentError.linesIterator
+              .foldLeft(List.empty[List[String]]) {
+                case (curr :: rest, elem) if elem != firstLine =>
+                  (elem :: curr) :: rest
+                case (acc, elem) => List(elem) :: acc
+              }
+              .map(_.reverse.mkString("\n"))
+              .distinct
+              .reverse
+              .mkString("\n")
+          }
+          .getOrElse(currentError)
+      }
+      if (previousError.contains(deduplicatedError)) None
+      else Some(deduplicatedError)
+    }
+  }
+  def append(message: String): CurrentAndPreviousError = {
+    val newCurrent =
+      if (currentError.isEmpty()) message else currentError + s"\n$message"
+    new CurrentAndPreviousError(previousError, newCurrent)
+  }
+  def isEmpty: Boolean = currentError.isEmpty()
+  def newError(): CurrentAndPreviousError = {
+    error match {
```

Comment:
Looks like error might be None if it has been deduplicated? Right? Could we do it more explicitly here in `newError`? I would opt to inline the logic from `error` here and just update the error var?

Reply:
I don't really understand this comment. I refactored the deduplication code so I hope it's more readable now.